sw/device/lib/aes.h

```diff
@@ -0,0 +1,112 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef _F_LIB_AES_H__
+#define _F_LIB_AES_H__
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/**
+ * Supported AES modes: encode or decode.
+ */
+typedef enum aes_mode { kAesEnc = 0, kAesDec = 1 } aes_mode_t;
+
+/**
+ * Supported AES key lengths: 128 bit, 192 bit or 256 bit. The hardware uses a
+ * one-hot encoding.
+ */
+typedef enum aes_key_len {
+  kAes128 = 0x1,
+  kAes192 = 0x2,
+  kAes256 = 0x4
+} aes_key_len_t;
+
+/**
+ * AES unit configuration options.
+ */
+typedef struct aes_cfg {
+  /** Operational mode @see aes_mode. */
+  aes_mode_t mode;
+  /** Key length @see aes_key_len. */
+  aes_key_len_t key_len;
+  /** Set to 1 to only start upon getting a trigger signal. */
+  bool manual_start_trigger;
+  /** Set to 1 to not stall when previous output data has not been read. */
+  bool force_data_overwrite;
+} aes_cfg_t;
+
+/**
+ * Intialize AES unit to desired mode.
+ *
+ * @param aes_cfg AES configuration settings.
+ */
+void aes_init(aes_cfg_t aes_cfg);
+
+/**
+ * Pass initial encryption key to AES unit.
+ *
+ * @param key              pointer to key.
+ * @param key_len_in_bytes key length in bytes (16, 24, 32)
+ */
+void aes_key_put(const void *key, size_t key_len_in_bytes);
+
+/**
+ * Wait for AES unit to be ready for new input data and then
+ * pass one 16B block of input data to AES unit.
+ *
+ * @param data pointer to input buffer.
+ */
+void aes_data_put_wait(const void *data);
+
+/**
+ * Pass one 16B block of input data to AES unit.
+ *
+ * @param data pointer to input buffer.
+ */
+void aes_data_put(const void *data);
+
+/**
+ * Wait for AES unit to have valid output data and then
+ * get one 16B block of output data from AES unit.
+ *
+ * @param data pointer to output buffer.
+ */
+void aes_data_get_wait(void *data);
+
+/**
+ * Get one 16B block of output data from AES unit.
+ *
+ * @param data pointer to output buffer.
+ */
+void aes_data_get(void *data);
+
+/**
+ * Check AES unit for being ready to accept new input data.
+ *
+ * @return true if ready for new input data, false otherwise.
+ */
+bool aes_data_ready(void);
+
+/**
+ * Check AES unit for having valid output data.
+ *
+ * @return true if valid output data available, false otherwise.
+ */
+bool aes_data_valid(void);
```

Comment:
I find the name "aes_data_valid" and the previous "aes_data_ready" a tad ambiguous, you can't tell from looking at them which one is the input and which one is the output.  ("get" and "put" even higher above don't really have that problem because there is an implied direction of data flow, but I would suspect that if you start relabeling valid/ready you'll want to make all of the function names have consistent prefixes.)

Reply:
I could rename them to `aes_data_get_valid()` and `aes_data_put_ready()` for example. But I think we can also merge now. This will have anyway have to be reworked once the DIF discussions have settled.

Reply:
This is a good point, I think it is fine to resolve this in the DIF.  (This was more of an informative comment anyway, which is why I approved rather than requesting changes.)